File: sdk/php/src/CoreOperators/XmlOperator.php
<?php

namespace TuskLang\CoreOperators;

use TuskLang\CoreOperators\BaseOperator;

/**
 * XML Operator for parsing, generating, and manipulating XML data
 */
class XmlOperator extends BaseOperator
{
    public function execute(array $config, array $context = []): mixed
    {
        $action = $config['action'] ?? 'parse';
        $data = $config['data'] ?? '';
        $options = $config['options'] ?? [];

        try {
            switch ($action) {
                case 'parse':
                    return $this->parseXml($data, $options);
                case 'generate':
                    return $this->generateXml($data, $options);
                case 'validate':
                    return $this->validateXml($data, $options);
                case 'transform':
                    return $this->transformXml($data, $options);
                case 'query':
                    return $this->queryXml($data, $options);
                case 'merge':
                    return $this->mergeXml($data, $options);
                default:
                    throw new \Exception("Unknown XML action: $action");
            }
        } catch (\Exception $e) {
            error_log("XML Operator error: " . $e->getMessage());
            throw $e;
        }
    }

    /**
     * Parse XML string to array
     */
    private function parseXml(string $xmlString, array $options): array
    {
        $flags = $options['flags'] ?? LIBXML_NOCDATA | LIBXML_NOERROR;
        $encoding = $options['encoding'] ?? 'UTF-8';
        
        libxml_use_internal_errors(true);
        $xml = simplexml_load_string($xmlString, 'SimpleXMLElement', $flags);
        
        if ($xml === false) {
            $errors = libxml_get_errors();
            libxml_clear_errors();
            throw new \Exception("XML parsing failed: " . implode(', ', array_map(fn($e) => $e->message, $errors)));
        }

        return $this->xmlToArray($xml);
    }

    /**
     * Generate XML from array
     */
    private function generateXml(array $data, array $options): string
    {
        $rootElement = $options['root'] ?? 'root';
        $version = $options['version'] ?? '1.0';
        $encoding = $options['encoding'] ?? 'UTF-8';
        $indent = $options['indent'] ?? true;

        $xml = new \XMLWriter();
        $xml->openMemory();
        $xml->setIndent($indent);
        $xml->startDocument($version, $encoding);
        $xml->startElement($rootElement);

        $this->arrayToXml($xml, $data);
        
        $xml->endElement();
        return $xml->outputMemory();
    }

    /**
     * Validate XML against schema
     */
    private function validateXml(string $xmlString, array $options): array
    {
        $schema = $options['schema'] ?? null;
        $dtd = $options['dtd'] ?? null;

        libxml_use_internal_errors(true);
        
        if ($schema) {
            $dom = new \DOMDocument();
            $dom->loadXML($xmlString);
            $valid = $dom->schemaValidate($schema);
        } elseif ($dtd) {
            $dom = new \DOMDocument();
            $dom->loadXML($xmlString);
            $valid = $dom->validate();
        } else {
            // Basic well-formedness check
            $valid = simplexml_load_string($xmlString) !== false;
        }

        $errors = libxml_get_errors();
        libxml_clear_errors();

        return [
            'valid' => $valid,
            'errors' => array_map(fn($e) => $e->message, $errors)
        ];
    }

    /**
     * Transform XML using XSLT
     */
    private function transformXml(string $xmlString, array $options): string
    {
        $xslt = $options['xslt'] ?? '';
        if (empty($xslt)) {
            throw new \Exception("XSLT stylesheet required for transformation");
        }

        $xmlDoc = new \DOMDocument();
        $xmlDoc->loadXML($xmlString);

        $xsltDoc = new \DOMDocument();
        $xsltDoc->loadXML($xslt);

        $processor = new \XSLTProcessor();
        $processor->importStylesheet($xsltDoc);

        return $processor->transformToXml($xmlDoc);
    }

    /**
     * Query XML using XPath
     */
    private function queryXml(string $xmlString, array $options): array
    {
        $xpath = $options['xpath'] ?? '';
        if (empty($xpath)) {
            throw new \Exception("XPath expression required for querying");
        }

        $dom = new \DOMDocument();
        $dom->loadXML($xmlString);
        
        $xpathObj = new \DOMXPath($dom);
        $nodes = $xpathObj->query($xpath);

        $results = [];
        foreach ($nodes as $node) {
            $results[] = $this->nodeToArray($node);
        }

        return $results;
    }

    /**
     * Merge multiple XML documents
     */
    private function mergeXml(array $xmlStrings, array $options): string
    {
        $rootElement = $options['root'] ?? 'merged';
        $strategy = $options['strategy'] ?? 'append'; // append, merge, replace

        $xml = new \XMLWriter();
        $xml->openMemory();
        $xml->setIndent(true);
        $xml->startDocument('1.0', 'UTF-8');
        $xml->startElement($rootElement);

        foreach ($xmlStrings as $xmlString) {
            $dom = new \DOMDocument();
            $dom->loadXML($xmlString);
            
            $this->importNode($xml, $dom->documentElement, $strategy);
        }

        $xml->endElement();
        return $xml->outputMemory();
    }

    /**
     * Convert SimpleXMLElement to array
     */
    private function xmlToArray(\SimpleXMLElement $xml): array
    {
        $array = [];
        
        foreach ($xml->children() as $child) {
            $name = $child->getName();
            $value = (string)$child;
            
            if ($child->count() > 0) {
                $array[$name] = $this->xmlToArray($child);
            } else {
                $array[$name] = $value;
            }
        }

        return $array;
    }

    /**
     * Convert array to XML
     */
    private function arrayToXml(\XMLWriter $xml, array $data): void
    {
        foreach ($data as $key => $value) {
            if (is_array($value)) {
                $xml->startElement($key);
                $this->arrayToXml($xml, $value);
                $xml->endElement();
            } else {
                $xml->writeElement($key, $value);
            }
        }
    }

    /**
     * Convert DOMNode to array
     */
    private function nodeToArray(\DOMNode $node): array
    {
        $array = [];
        
        if ($node->nodeType === XML_ELEMENT_NODE) {
            $array['name'] = $node->nodeName;
            $array['value'] = $node->textContent;
            
            if ($node->hasAttributes()) {
                $array['attributes'] = [];
                foreach ($node->attributes as $attr) {
                    $array['attributes'][$attr->name] = $attr->value;
                }
            }
            
            if ($node->hasChildNodes()) {
                $array['children'] = [];
                foreach ($node->childNodes as $child) {
                    if ($child->nodeType === XML_ELEMENT_NODE) {
                        $array['children'][] = $this->nodeToArray($child);
                    }
                }
            }
        }

        return $array;
    }

    /**
     * Import DOMNode into XMLWriter
     */
    private function importNode(\XMLWriter $xml, \DOMNode $node, string $strategy): void
    {
        if ($node->nodeType === XML_ELEMENT_NODE) {
            $xml->startElement($node->nodeName);
            
            if ($node->hasAttributes()) {
                foreach ($node->attributes as $attr) {
                    $xml->writeAttribute($attr->name, $attr->value);
                }
            }
            
            if ($node->hasChildNodes()) {
                foreach ($node->childNodes as $child) {
                    if ($child->nodeType === XML_ELEMENT_NODE) {
                        $this->importNode($xml, $child, $strategy);
                    } elseif ($child->nodeType === XML_TEXT_NODE) {
                        $xml->text($child->textContent);
                    }
                }
            }
            
            $xml->endElement();
        }
    }
} 

File: sdk/csharp/csharp-sdk-package-20250721_013844/operators/Operators/DataProcessing/XmlOperator.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;
using System.Linq; // Added for .Any()

namespace TuskLang.Operators.DataProcessing
{
    /// <summary>
    /// XML Operator for TuskLang C# SDK
    /// 
    /// Provides XML processing capabilities with support for:
    /// - XML parsing and serialization
    /// - XPath queries and navigation
    /// - XML validation and formatting
    /// - XML transformation (XSLT)
    /// - XML schema validation
    /// - XML manipulation and editing
    /// 
    /// Usage:
    /// ```csharp
    /// // Parse XML
    /// var result = @xml({
    ///   action: "parse",
    ///   data: "<root><item>value</item></root>"
    /// })
    /// 
    /// // XPath query
    /// var result = @xml({
    ///   action: "query",
    ///   data: xmlData,
    ///   xpath: "//item[@id='1']"
    /// })
    /// ```
    /// </summary>
    public class XmlOperator : BaseOperator
    {
        public XmlOperator()
        {
            Version = "2.0.0";
            RequiredFields = new List<string> { "action" };
            OptionalFields = new List<string> 
            { 
                "data", "xpath", "schema", "format", "indent", "validate", 
                "transform", "merge", "diff", "patch", "timeout", "encoding" 
            };
            
            DefaultConfig = new Dictionary<string, object>
            {
                ["format"] = "pretty",
                ["indent"] = 2,
                ["validate"] = true,
                ["timeout"] = 300,
                ["encoding"] = "UTF-8"
            };
        }
        
        public override string GetName() => "xml";
        
        protected override string GetDescription() => "XML processing operator for parsing, querying, and manipulating XML data";
        
        protected override Dictionary<string, string> GetExamples()
        {
            return new Dictionary<string, string>
            {
                ["parse"] = "@xml({action: \"parse\", data: \"<root><item>value</item></root>\"})",
                ["stringify"] = "@xml({action: \"stringify\", data: xmlObject})",
                ["query"] = "@xml({action: \"query\", data: xmlData, xpath: \"//item[@id='1']\"})",
                ["validate"] = "@xml({action: \"validate\", data: xmlData, schema: schemaData})"
            };
        }
        
        protected override Dictionary<string, string> GetErrorCodes()
        {
            return new Dictionary<string, string>
            {
                ["INVALID_ACTION"] = "Invalid XML action",
                ["INVALID_XML"] = "Invalid XML data",
                ["PARSE_ERROR"] = "XML parsing error",
                ["XPATH_ERROR"] = "XPath query error",
                ["SCHEMA_VALIDATION_ERROR"] = "XML schema validation failed",
                ["TRANSFORM_ERROR"] = "XML transformation error",
                ["MERGE_ERROR"] = "XML merge error",
                ["TIMEOUT_EXCEEDED"] = "XML operation timeout exceeded"
            };
        }
        
        protected override async Task<object> ExecuteOperatorAsync(Dictionary<string, object> config, Dictionary<string, object> context)
        {
            var action = GetContextValue<string>(config, "action", "");
            var data = ResolveVariable(config.GetValueOrDefault("data"), context);
            var xpath = GetContextValue<string>(config, "xpath", "");
            var schema = ResolveVariable(config.GetValueOrDefault("schema"), context);
            var format = GetContextValue<string>(config, "format", "pretty");
            var indent = GetContextValue<int>(config, "indent", 2);
            var validate = GetContextValue<bool>(config, "validate", true);
            var transform = GetContextValue<string>(config, "transform", "");
            var merge = ResolveVariable(config.GetValueOrDefault("merge"), context);
            var diff = ResolveVariable(config.GetValueOrDefault("diff"), context);
            var patch = ResolveVariable(config.GetValueOrDefault("patch"), context);
            var timeout = GetContextValue<int>(config, "timeout", 300);
            var encoding = GetContextValue<string>(config, "encoding", "UTF-8");
            
            if (string.IsNullOrEmpty(action))
                throw new ArgumentException("Action is required");
            
            var startTime = DateTime.UtcNow;
            
            try
            {
                // Check timeout
                if ((DateTime.UtcNow - startTime).TotalSeconds > timeout)
                {
                    throw new TimeoutException("XML operation timeout exceeded");
                }
                
                switch (action.ToLower())
                {
                    case "parse":
                        return await ParseXmlAsync(data, validate, encoding);
                    
                    case "stringify":
                        return await StringifyXmlAsync(data, format, indent, encoding);
                    
                    case "query":
                        return await QueryXmlAsync(data, xpath);
                    
                    case "validate":
                        return await ValidateXmlAsync(data, schema);
                    
                    case "transform":
                        return await TransformXmlAsync(data, transform);
                    
                    case "merge":
                        return await MergeXmlAsync(data, merge);
                    
                    case "diff":
                        return await DiffXmlAsync(data, diff);
                    
                    case "patch":
                        return await PatchXmlAsync(data, patch);
                    
                    case "format":
                        return await FormatXmlAsync(data, format, indent, encoding);
                    
                    case "minify":
                        return await MinifyXmlAsync(data, encoding);
                    
                    case "flatten":
                        return await FlattenXmlAsync(data);
                    
                    case "unflatten":
                        return await UnflattenXmlAsync(data);
                    
                    default:
                        throw new ArgumentException($"Unknown XML action: {action}");
                }
            }
            catch (Exception ex)
            {
                Log("error", "XML operation failed", new Dictionary<string, object>
                {
                    ["action"] = action,
                    ["error"] = ex.Message
                });
                
                return new Dictionary<string, object>
                {
                    ["success"] = false,
                    ["error"] = ex.Message,
                    ["action"] = action
                };
            }
        }
        
        /// <summary>
        /// Parse XML string to object
        /// </summary>
        private async Task<object> ParseXmlAsync(object data, bool validate, string encoding)
        {
            if (data == null)
                throw new ArgumentException("Data is required for parsing");
            
            var xmlString = data.ToString();
            
            try
            {
                var doc = XDocument.Parse(xmlString);
                var result = ConvertXDocumentToDictionary(doc);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["parsed"] = result,
                    ["valid"] = true,
                    ["encoding"] = encoding
                };
            }
            catch (XmlException ex)
            {
                if (validate)
                {
                    throw new ArgumentException($"Invalid XML: {ex.Message}");
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = false,
                    ["error"] = ex.Message,
                    ["valid"] = false
                };
            }
        }
        
        /// <summary>
        /// Convert object to XML string
        /// </summary>
        private async Task<object> StringifyXmlAsync(object data, string format, int indent, string encoding)
        {
            if (data == null)
                throw new ArgumentException("Data is required for stringification");
            
            try
            {
                var doc = ConvertDictionaryToXDocument(data);
                var options = new SaveOptions();
                
                if (format == "pretty")
                {
                    options = SaveOptions.None;
                }
                else
                {
                    options = SaveOptions.DisableFormatting;
                }
                
                var xmlString = doc.ToString(options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["stringified"] = xmlString,
                    ["format"] = format,
                    ["encoding"] = encoding
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Stringification failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Query XML with XPath
        /// </summary>
        private async Task<object> QueryXmlAsync(object data, string xpath)
        {
            if (data == null)
                throw new ArgumentException("Data is required for querying");
            
            if (string.IsNullOrEmpty(xpath))
                throw new ArgumentException("XPath is required for querying");
            
            try
            {
                XDocument doc;
                if (data is string xmlString)
                {
                    doc = XDocument.Parse(xmlString);
                }
                else if (data is XDocument xdoc)
                {
                    doc = xdoc;
                }
                else
                {
                    doc = ConvertDictionaryToXDocument(data);
                }
                
                var results = new List<object>();
                var navigator = doc.CreateNavigator();
                var iterator = navigator.Select(xpath);
                
                while (iterator.MoveNext())
                {
                    var node = iterator.Current;
                    if (node != null)
                    {
                        results.Add(node.OuterXml);
                    }
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["results"] = results,
                    ["count"] = results.Count,
                    ["xpath"] = xpath
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"XPath query failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Validate XML against schema
        /// </summary>
        private async Task<object> ValidateXmlAsync(object data, object schema)
        {
            if (data == null)
                throw new ArgumentException("Data is required for validation");
            
            // Simplified validation
            // In a real implementation, you would use a proper XML schema validator
            var isValid = true;
            var errors = new List<string>();
            
            try
            {
                XDocument doc;
                if (data is string xmlString)
                {
                    doc = XDocument.Parse(xmlString);
                }
                else if (data is XDocument xdoc)
                {
                    doc = xdoc;
                }
                else
                {
                    doc = ConvertDictionaryToXDocument(data);
                }
                
                // Basic validation - check if XML is well-formed
                if (doc.Root == null)
                {
                    isValid = false;
                    errors.Add("XML document has no root element");
                }
            }
            catch (Exception ex)
            {
                isValid = false;
                errors.Add($"XML validation failed: {ex.Message}");
            }
            
            return new Dictionary<string, object>
            {
                ["success"] = true,
                ["valid"] = isValid,
                ["errors"] = errors,
                ["error_count"] = errors.Count
            };
        }
        
        /// <summary>
        /// Transform XML data
        /// </summary>
        private async Task<object> TransformXmlAsync(object data, string transform)
        {
            if (data == null)
                throw new ArgumentException("Data is required for transformation");
            
            if (string.IsNullOrEmpty(transform))
                throw new ArgumentException("Transform expression is required");
            
            try
            {
                // Simplified transformation
                // In a real implementation, you would parse and execute the transform expression
                var transformed = data; // Placeholder for actual transformation
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["transformed"] = transformed,
                    ["transform"] = transform
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Transformation failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Merge XML objects
        /// </summary>
        private async Task<object> MergeXmlAsync(object data, object merge)
        {
            if (data == null)
                throw new ArgumentException("Data is required for merging");
            
            if (merge == null)
                throw new ArgumentException("Merge data is required");
            
            try
            {
                // Simplified merge
                // In a real implementation, you would implement deep merge logic
                var merged = new Dictionary<string, object>();
                
                if (data is Dictionary<string, object> dataDict)
                {
                    foreach (var kvp in dataDict)
                    {
                        merged[kvp.Key] = kvp.Value;
                    }
                }
                
                if (merge is Dictionary<string, object> mergeDict)
                {
                    foreach (var kvp in mergeDict)
                    {
                        merged[kvp.Key] = kvp.Value;
                    }
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["merged"] = merged
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Merge failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Diff XML objects
        /// </summary>
        private async Task<object> DiffXmlAsync(object data, object diff)
        {
            if (data == null)
                throw new ArgumentException("Data is required for diffing");
            
            if (diff == null)
                throw new ArgumentException("Diff data is required");
            
            try
            {
                // Simplified diff
                // In a real implementation, you would implement proper diff logic
                var differences = new List<object>();
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["differences"] = differences,
                    ["has_differences"] = differences.Count > 0
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Diff failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Patch XML object
        /// </summary>
        private async Task<object> PatchXmlAsync(object data, object patch)
        {
            if (data == null)
                throw new ArgumentException("Data is required for patching");
            
            if (patch == null)
                throw new ArgumentException("Patch data is required");
            
            try
            {
                // Simplified patch
                // In a real implementation, you would implement proper patch logic
                var patched = data; // Placeholder for actual patching
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["patched"] = patched
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Patch failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Format XML
        /// </summary>
        private async Task<object> FormatXmlAsync(object data, string format, int indent, string encoding)
        {
            if (data == null)
                throw new ArgumentException("Data is required for formatting");
            
            try
            {
                XDocument doc;
                if (data is string xmlString)
                {
                    doc = XDocument.Parse(xmlString);
                }
                else if (data is XDocument xdoc)
                {
                    doc = xdoc;
                }
                else
                {
                    doc = ConvertDictionaryToXDocument(data);
                }
                
                var options = format == "pretty" ? SaveOptions.None : SaveOptions.DisableFormatting;
                var formatted = doc.ToString(options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["formatted"] = formatted,
                    ["format"] = format,
                    ["encoding"] = encoding
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Formatting failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Minify XML
        /// </summary>
        private async Task<object> MinifyXmlAsync(object data, string encoding)
        {
            if (data == null)
                throw new ArgumentException("Data is required for minification");
            
            try
            {
                XDocument doc;
                if (data is string xmlString)
                {
                    doc = XDocument.Parse(xmlString);
                }
                else if (data is XDocument xdoc)
                {
                    doc = xdoc;
                }
                else
                {
                    doc = ConvertDictionaryToXDocument(data);
                }
                
                var minified = doc.ToString(SaveOptions.DisableFormatting);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["minified"] = minified,
                    ["encoding"] = encoding
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Minification failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Flatten XML object
        /// </summary>
        private async Task<object> FlattenXmlAsync(object data)
        {
            if (data == null)
                throw new ArgumentException("Data is required for flattening");
            
            try
            {
                var flattened = FlattenObject(data, "");
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["flattened"] = flattened
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Flattening failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Unflatten XML object
        /// </summary>
        private async Task<object> UnflattenXmlAsync(object data)
        {
            if (data == null)
                throw new ArgumentException("Data is required for unflattening");
            
            try
            {
                var unflattened = UnflattenObject(data);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["unflattened"] = unflattened
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Unflattening failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Convert XDocument to Dictionary
        /// </summary>
        private Dictionary<string, object> ConvertXDocumentToDictionary(XDocument doc)
        {
            var result = new Dictionary<string, object>();
            
            if (doc.Root != null)
            {
                result[doc.Root.Name.LocalName] = ConvertXElementToDictionary(doc.Root);
            }
            
            return result;
        }
        
        /// <summary>
        /// Convert XElement to Dictionary
        /// </summary>
        private object ConvertXElementToDictionary(XElement element)
        {
            var result = new Dictionary<string, object>();
            
            // Add attributes
            foreach (var attr in element.Attributes())
            {
                result[$"@{attr.Name.LocalName}"] = attr.Value;
            }
            
            // Add child elements
            var children = element.Elements();
            if (children.Any())
            {
                foreach (var child in children)
                {
                    var childName = child.Name.LocalName;
                    var childValue = ConvertXElementToDictionary(child);
                    
                    if (result.ContainsKey(childName))
                    {
                        // Handle multiple elements with same name
                        if (result[childName] is List<object> list)
                        {
                            list.Add(childValue);
                        }
                        else
                        {
                            result[childName] = new List<object> { result[childName], childValue };
                        }
                    }
                    else
                    {
                        result[childName] = childValue;
                    }
                }
            }
            else
            {
                // Leaf element with text content
                result["#text"] = element.Value;
            }
            
            return result;
        }
        
        /// <summary>
        /// Convert Dictionary to XDocument
        /// </summary>
        private XDocument ConvertDictionaryToXDocument(object data)
        {
            if (data is Dictionary<string, object> dict)
            {
                var rootName = dict.Keys.FirstOrDefault();
                if (!string.IsNullOrEmpty(rootName))
                {
                    var rootElement = new XElement(rootName);
                    var rootData = dict[rootName];
                    
                    if (rootData is Dictionary<string, object> rootDict)
                    {
                        PopulateXElement(rootElement, rootDict);
                    }
                    
                    return new XDocument(rootElement);
                }
            }
            
            return new XDocument(new XElement("root"));
        }
        
        /// <summary>
        /// Populate XElement from Dictionary
        /// </summary>
        private void PopulateXElement(XElement element, Dictionary<string, object> data)
        {
            foreach (var kvp in data)
            {
                if (kvp.Key.StartsWith("@"))
                {
                    // Attribute
                    var attrName = kvp.Key.Substring(1);
                    element.SetAttributeValue(attrName, kvp.Value);
                }
                else if (kvp.Key == "#text")
                {
                    // Text content
                    element.Value = kvp.Value?.ToString() ?? "";
                }
                else
                {
                    // Child element
                    if (kvp.Value is List<object> list)
                    {
                        foreach (var item in list)
                        {
                            var childElement = new XElement(kvp.Key);
                            if (item is Dictionary<string, object> childDict)
                            {
                                PopulateXElement(childElement, childDict);
                            }
                            else
                            {
                                childElement.Value = item?.ToString() ?? "";
                            }
                            element.Add(childElement);
                        }
                    }
                    else if (kvp.Value is Dictionary<string, object> childDict)
                    {
                        var childElement = new XElement(kvp.Key);
                        PopulateXElement(childElement, childDict);
                        element.Add(childElement);
                    }
                    else
                    {
                        var childElement = new XElement(kvp.Key, kvp.Value);
                        element.Add(childElement);
                    }
                }
            }
        }
        
        /// <summary>
        /// Flatten object to key-value pairs
        /// </summary>
        private Dictionary<string, object> FlattenObject(object data, string prefix)
        {
            var result = new Dictionary<string, object>();
            
            if (data is Dictionary<string, object> dict)
            {
                foreach (var kvp in dict)
                {
                    var key = string.IsNullOrEmpty(prefix) ? kvp.Key : $"{prefix}.{kvp.Key}";
                    
                    if (kvp.Value is Dictionary<string, object> nestedDict)
                    {
                        var nested = FlattenObject(kvp.Value, key);
                        foreach (var nestedKvp in nested)
                        {
                            result[nestedKvp.Key] = nestedKvp.Value;
                        }
                    }
                    else
                    {
                        result[key] = kvp.Value;
                    }
                }
            }
            
            return result;
        }
        
        /// <summary>
        /// Unflatten object from key-value pairs
        /// </summary>
        private object UnflattenObject(object data)
        {
            if (data is Dictionary<string, object> dict)
            {
                var result = new Dictionary<string, object>();
                
                foreach (var kvp in dict)
                {
                    var parts = kvp.Key.Split('.');
                    var current = result;
                    
                    for (int i = 0; i < parts.Length - 1; i++)
                    {
                        var part = parts[i];
                        if (!current.ContainsKey(part))
                        {
                            current[part] = new Dictionary<string, object>();
                        }
                        current = (Dictionary<string, object>)current[part];
                    }
                    
                    current[parts[parts.Length - 1]] = kvp.Value;
                }
                
                return result;
            }
            
            return data;
        }
        
        protected override ValidationResult CustomValidate(Dictionary<string, object> config)
        {
            var result = new ValidationResult();
            
            if (!config.ContainsKey("action"))
            {
                result.Errors.Add("Action is required");
            }
            
            var action = GetContextValue<string>(config, "action", "");
            var validActions = new[] { "parse", "stringify", "query", "validate", "transform", "merge", "diff", "patch", "format", "minify", "flatten", "unflatten" };
            
            if (!string.IsNullOrEmpty(action) && !Array.Exists(validActions, a => a == action.ToLower()))
            {
                result.Errors.Add($"Invalid action: {action}. Valid actions are: {string.Join(", ", validActions)}");
            }
            
            if (config.TryGetValue("timeout", out var timeout) && timeout is int timeoutValue && timeoutValue <= 0)
            {
                result.Errors.Add("Timeout must be positive");
            }
            
            if (config.TryGetValue("indent", out var indent) && indent is int indentValue && indentValue < 0)
            {
                result.Errors.Add("Indent must be non-negative");
            }
            
            return result;
        }
    }
} 